File: src/main/python/vulnerable_app.py
"""
VULNERABLE CODE - DO NOT USE IN PRODUCTION
Sample vulnerable Python code for SAST testing
CWE vulnerabilities included for security scanning
"""

import sqlite3
import pickle
import os
import subprocess
import hashlib
from flask import Flask, request

app = Flask(__name__)

# Vulnerability: Hardcoded Credentials (CWE-798)
DB_HOST = "localhost"
DB_USER = "admin"
DB_PASSWORD = "SuperSecret123!"
API_KEY = "sk-1234567890abcdefghijklmnop"
SECRET_TOKEN = "mysecrettoken12345"

# Database connection
db_connection = None


@app.route('/search')
def search_user():
    """
    Vulnerability: SQL Injection (CWE-89)
    User input is directly concatenated into SQL query
    """
    username = request.args.get('username')
    
    query = f"SELECT * FROM users WHERE username = '{username}'"
    cursor = db_connection.cursor()
    cursor.execute(query)  # SQL Injection vulnerability
    
    result = cursor.fetchone()
    return {"user": result}


@app.route('/execute', methods=['POST'])
def execute_command():
    """
    Vulnerability: Command Injection (CWE-78)
    User input passed directly to shell execution
    """
    command = request.json.get('command')
    
    # Command Injection vulnerability
    result = os.system(command)
    return {"status": "executed"}


@app.route('/file')
def read_file():
    """
    Vulnerability: Path Traversal (CWE-22)
    User input used directly in file path without validation
    """
    filename = request.args.get('filename')
    
    base_path = "/uploads/"
    file_path = base_path + filename  # No validation - allows ../ traversal
    
    try:
        with open(file_path, 'r') as f:
            content = f.read()
        return {"content": content}
    except Exception as e:
        return {"error": str(e)}


@app.route('/deserialize', methods=['POST'])
def deserialize_object():
    """
    Vulnerability: Insecure Deserialization (CWE-502)
    Using pickle to deserialize untrusted data
    """
    data = request.files['file'].read()
    
    try:
        obj = pickle.loads(data)  # Insecure deserialization
        return {"object": str(obj)}
    except Exception as e:
        return {"error": str(e)}


@app.route('/hash')
def hash_password():
    """
    Vulnerability: Weak Cryptography (CWE-327)
    Using weak hashing algorithm without salt
    """
    password = request.args.get('password')
    
    # MD5 is cryptographically broken - insecure
    hash_obj = hashlib.md5(password.encode())
    hashed = hash_obj.hexdigest()
    
    return {"hash": hashed}


@app.route('/user/<user_id>')
def get_user_data(user_id):
    """
    Vulnerability: Sensitive Data Exposure (CWE-200)
    Returns sensitive user data without proper authorization
    """
    query = f"SELECT * FROM users WHERE id = {user_id}"
    cursor = db_connection.cursor()
    cursor.execute(query)
    
    result = cursor.fetchone()
    
    # Returning sensitive data without filtering
    return {
        "id": result[0],
        "username": result[1],
        "email": result[2],
        "ssn": result[3],  # Sensitive
        "credit_card": result[4],  # Sensitive
        "password_hash": result[5]  # Should not be exposed
    }


@app.route('/upload', methods=['POST'])
def upload_file():
    """
    Vulnerability: Unrestricted File Upload (CWE-434)
    No validation of uploaded file type or content
    """
    file = request.files['file']
    
    # No file type validation - allows arbitrary file uploads
    filename = file.filename
    file.save(f'/uploads/{filename}')
    
    return {"status": "uploaded", "filename": filename}


@app.route('/execute-subprocess')
def execute_subprocess():
    """
    Vulnerability: Unsafe Subprocess Usage
    User input used in subprocess without proper escaping
    """
    cmd_arg = request.args.get('arg')
    
    # Dangerous: shell=True with user input
    result = subprocess.run(f"echo {cmd_arg}", shell=True, capture_output=True)
    
    return {"output": result.stdout.decode()}


def debug_eval():
    """
    Vulnerability: Code Injection (CWE-95)
    Using eval() with user-controlled input
    """
    user_input = request.args.get('code')
    
    try:
        result = eval(user_input)  # Dangerous code injection
        return {"result": str(result)}
    except Exception as e:
        return {"error": str(e)}


def check_password(provided, stored):
    """
    Vulnerability: Weak Password Verification
    Timing attack vulnerability - no constant-time comparison
    """
    if provided == stored:  # Not constant-time, vulnerable to timing attacks
        return True
    return False


if __name__ == '__main__':
    # Vulnerability: Running Flask in debug mode in production
    app.run(debug=True, host='0.0.0.0')  # Debug=True is insecure

File: src/main/javascript/vulnerable-app.js
/**
 * VULNERABLE CODE - DO NOT USE IN PRODUCTION
 * Sample vulnerable Node.js/JavaScript code for SAST testing
 */

const express = require('express');
const sqlite3 = require('sqlite3').verbose();
const mysql = require('mysql');
const eval = require('eval');
const path = require('path');
const fs = require('fs');

const app = express();

// Vulnerability: Hardcoded Credentials (CWE-798)
const DB_HOST = 'localhost';
const DB_USER = 'admin';
const DB_PASSWORD = 'SuperSecret123!';
const API_KEY = 'sk-1234567890abcdefghijklmnop';
const JWT_SECRET = 'mysecretkey123456';

/**
 * Vulnerability: SQL Injection (CWE-89)
 * User input directly concatenated into SQL query
 */
app.get('/api/users/search', (req, res) => {
    const username = req.query.username;
    
    // SQL Injection vulnerability
    const query = `SELECT * FROM users WHERE username = '${username}'`;
    
    db.all(query, (err, rows) => {
        if (err) {
            return res.status(500).json({ error: err.message });
        }
        res.json({ users: rows });
    });
});

/**
 * Vulnerability: Cross-Site Scripting (XSS) (CWE-79)
 * User input reflected without sanitization
 */
app.get('/api/search-results', (req, res) => {
    const searchTerm = req.query.q;
    
    // Reflected XSS vulnerability
    const html = `
        <html>
            <body>
                <h1>Search Results for: ${searchTerm}</h1>
                <p>Results found...</p>
            </body>
        </html>
    `;
    
    res.send(html);
});

/**
 * Vulnerability: Command Injection (CWE-78)
 * User input passed to shell commands
 */
app.post('/api/execute', (req, res) => {
    const command = req.body.command;
    
    // Command Injection vulnerability
    const child_process = require('child_process');
    child_process.exec(command, (error, stdout, stderr) => {
        if (error) {
            return res.status(500).json({ error: error.message });
        }
        res.json({ output: stdout });
    });
});

/**
 * Vulnerability: Path Traversal (CWE-22)
 * User input used directly in file path
 */
app.get('/api/file', (req, res) => {
    const filename = req.query.filename;
    
    // Path Traversal vulnerability - no validation
    const filePath = path.join(__dirname, '/uploads/', filename);
    
    fs.readFile(filePath, 'utf8', (err, data) => {
        if (err) {
            return res.status(500).json({ error: err.message });
        }
        res.json({ content: data });
    });
});

/**
 * Vulnerability: Insecure Deserialization (CWE-502)
 * Using eval() to deserialize untrusted data
 */
app.post('/api/deserialize', (req, res) => {
    const data = req.body.data;
    
    try {
        // eval() is extremely dangerous
        const obj = eval(`(${data})`);
        res.json({ object: obj });
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

/**
 * Vulnerability: Sensitive Data in Logs (CWE-532)
 * Logging sensitive information
 */
app.post('/api/login', (req, res) => {
    const username = req.body.username;
    const password = req.body.password;
    
    // Logging sensitive data
    console.log(`User login attempt: username=${username}, password=${password}`);
    console.log(`Credentials received: ${username}:${password}`);
    
    // Storing password in memory without hashing
    const user = { username, password };
    res.json({ success: true, user });
});

/**
 * Vulnerability: Sensitive Data Exposure (CWE-200)
 * Returning sensitive data without authorization
 */
app.get('/api/users/:id', (req, res) => {
    const userId = req.params.id;
    
    const query = `SELECT * FROM users WHERE id = ${userId}`;
    
    db.get(query, (err, row) => {
        if (err) {
            return res.status(500).json({ error: err.message });
        }
        
        // Returning all sensitive data
        res.json({
            id: row.id,
            username: row.username,
            email: row.email,
            ssn: row.ssn,
            creditCard: row.credit_card,
            passwordHash: row.password_hash,
            apiKey: row.api_key
        });
    });
});

/**
 * Vulnerability: Unrestricted File Upload (CWE-434)
 * No validation of uploaded files
 */
app.post('/api/upload', (req, res) => {
    const file = req.files.file;
    
    // No file type validation
    const uploadPath = path.join(__dirname, '/uploads/', file.name);
    
    file.mv(uploadPath, (err) => {
        if (err) {
            return res.status(500).json({ error: err.message });
        }
        res.json({ filename: file.name, path: uploadPath });
    });
});

/**
 * Vulnerability: Weak Cryptography (CWE-327)
 * Using MD5 for password hashing
 */
const crypto = require('crypto');

function hashPassword(password) {
    // MD5 is cryptographically broken
    return crypto.createHash('md5').update(password).digest('hex');
}

/**
 * Vulnerability: Insecure Random (CWE-338)
 * Using Math.random() for security purposes
 */
function generateToken() {
    // Math.random() is not cryptographically secure
    return Math.random().toString(36).substring(2, 15);
}

/**
 * Vulnerability: Cross-Site Request Forgery (CSRF) (CWE-352)
 * No CSRF token validation
 */
app.post('/api/transfer-money', (req, res) => {
    const amount = req.body.amount;
    const recipient = req.body.recipient;
    
    // No CSRF token check - vulnerable to CSRF attacks
    const query = `UPDATE accounts SET balance = balance - ${amount} WHERE user_id = ${req.userId}`;
    
    db.run(query, (err) => {
        if (err) {
            return res.status(500).json({ error: err.message });
        }
        res.json({ success: true });
    });
});

/**
 * Vulnerability: Insecure Direct Object Reference (IDOR) (CWE-639)
 * No authorization check
 */
app.get('/api/profile/:userId', (req, res) => {
    const userId = req.params.userId;
    
    // No check if the user is authorized to view this profile
    const query = `SELECT * FROM users WHERE id = ${userId}`;
    
    db.get(query, (err, row) => {
        if (err) {
            return res.status(500).json({ error: err.message });
        }
        res.json(row);
    });
});

// Vulnerability: Debug mode enabled
app.use((err, req, res, next) => {
    console.error('Full error details:', err); // Exposes stack trace
    res.status(500).json({
        error: err.message,
        stack: err.stack // Exposes stack trace to client
    });
});

// Database setup
const db = new sqlite3.Database('./vulnerable.db');

const PORT = process.env.PORT || 3000;
app.listen(PORT, '0.0.0.0', () => {
    console.log(`Server running on port ${PORT}`);
});
